### DISCUSSION TITLE
Trying to use a custom data source

### DISCUSSION THREAD

**Discussion by bigdogdman** [yajra/laravel-datatables#3147] [Q&A]
I have a custom data source, which is API endpoints. I am trying to use those endpoints (1 per page) as my "database" for my DataTables. I was told by a member of the DataTables community to use a custom ajax config setting for my DataTable, but I'm unsure on how to do that with Yajra. Is there anything I can do to make my DataTable not use the default db connection defined in .env? Here is my index.blade.php:

```
<x-app-layout>
    <x-slot name="slot">
        <div class="container text-white">
            <div class="card">
                <div class="card-header">REGIONS (***TESTING***)</div>
                <div class="card-body">
                    {{ $dataTable->table(['data' => $data]) }}
                </div>
            </div>
        </div>
        @push('scripts')
            {{ $dataTable->scripts(attributes: ['type' => 'module', 'serverSide' => true, 'ajax' => route('user-data')]) }}
        @endpush
    </x-slot>
</x-app-layout>

```

My 'user-data' route also returns a custom Eloquent Builder object that should include everything needed, but I can't get my DT to actually use it to see if it's correct or not.  Any help would be extremely appreciated! :) 

Update: Okay, I got my ajax callback route set up correctly (by reading [this page](https://github.com/yajra/laravel-datatables-docs/blob/10.0/html-builder-ajax.md)), but now it's not hitting it on initial load, whereas before it was (when it was throwing an error--and hitting the wrong endpoint), so no data loads, and the paginate buttons don't display. I have to manually change the number of records to get it to show any data at all. Also, is there a reason it's not using the initial $data object I'm passing to it to load onto the screen? Is there some way I can force it to do that? Thanks again! :)

**Comment by bigdogdman** [comment]
Further update: now the ajax callback isn't working anymore. It's hitting the callback, but it's acting like it's returning the wrong datatype again. This was working just a few days ago, and I haven't changed that much in the code; I even tried reverting my changes but I was still getting the error. The error I'm getting is `
Yajra\DataTables\Utilities\Helper::compileContent(): Argument #3 ($param) must be of type object|array, int given
` on my DataTables API endpoint return line: `return DataTables::of(collect($response->get('results')))->addColumn('action', 'regions.action')->make(true);`.
When I manipulate what I'm returning from my API endpoint, the error changes (depending on what's being returned), so I know it has something to do with what the DataTable is expecting back. However, when I return what _was_ working just last week, it gives me this setContent() error: `Symfony\Component\HttpFoundation\Response::setContent(): Argument #1 ($content) must be of type ?string, Yajra\DataTables\CollectionDataTable given`. 
I have no idea what I'm doing anymore....

Here is my endpoint code that, when I return `$returnData` it gives me the `compileContent()` error:
```
        $tokens = ['offset' => $queryParams['start'] ?? '0'];
        $response = Netbox::getData('dcim/regions?limit=0', $tokens);
        $data = $response->get('results');
        $returnData = DataTables::of($response)->addColumn('action', 'regions.action')->make(true);
        dd($returnData);
```
Netbox is my custom data model that I'm using in conjunction with my custom auth file that it's throwing the setContent() error on when I return what I _assume_ should be correct, which is `DataTables::of(collect($response->get('results')))`.

**Reply by bigdogdman** [reply]
Further update: I got the callback working again, but the DT is still not loading the data I'm passing to it on initial load. Isn't it supposed to do that? My index.blade is this:
```
<x-app-layout>
    <x-slot name="header">
        <x-footprint-menu />
    </x-slot>
    <x-slot name="slot">
        <div class="container text-white">
            <div class="card">
                <div class="card-header">{{ $title }}</div>
                <div class="card-body">
                    {{ $dataTable->table(['data' => $data]) }}
                </div>
            </div>
        </div>
        @push('scripts')
            {{ $dataTable->scripts(attributes: ['type' => 'module']) }}
        @endpush
    </x-slot>
</x-app-layout>
```
My htmlBuilder is this:
```
    public function html(): HtmlBuilder
    {
        return $this->builder()
                    ->setTableId('regions-table')
                    ->columns($this->getColumns())
                    ->minifiedAjax()
                    ->ajax(route('region-data'))
                    //*/
                    //->dom('Bfrtip')
                    ->orderBy([1, 'desc'])  // dynamic??
                    ->paging()
                    //->selectStyleSingle()
                    //->selectSelector('td:first-child')
                    /* ->buttons([
                        Button::make('excel'),
                        Button::make('csv'),
                        Button::make('pdf'),
                        Button::make('print'),
                        Button::make('reset'),
                        Button::make('reload')
                    ]) */;
                    /*/
                    ->drawCallbackWithLivewire() // not working for me...
                    ->orderBy(3, 'asc')
                    ->selectStyleOS()
                    ->selectSelector('td:first-child')
                    ->buttons([
                        Button::make(),
                        //CreateButton::make(),
                        //EditButton::make(),
                        //RemoveButton::make(),
                        //ReloadButton::make(),
                    ])
                    ->editors([         // this seems to break my DataTable
                        Editor::make()
                              ->fields([
                                  Field::make('video_id'),
                                  Text::make('title'),
                                  Select2::make('tags')
                                         ->data('tags[].name')
                                         ->label('Tags')
                                         ->opts([
                                             'multiple' => true,
                                         ])
                                         //->modelOptions(Netbox::class, 'name', 'name'),
                              ]),
                    ]);
                    //*/
    }
```
And my $data object I'm passing to index is this:
```
Yajra\DataTables\CollectionDataTable {#1618 ▼ // app/Http/Controllers/CustomController.php:190
  +request: 
Yajra\DataTables\Utilities\Request {#644 ▶}
  #logger: null
  #columns: array:377 [▼
    0 => 95
    1 => 98
    ...
	376 => 1564
  ]
  #columnDef: array:9 [▼
    "index" => true
    "ignore_getters" => false
    "append" => []
    "edit" => []
    "filter" => []
    "order" => []
    "only" => array:13 [▼
      0 => "id"
      1 => "url"
      2 => "display"
      3 => "name"
      4 => "slug"
      5 => "parent"
      6 => "description"
      7 => "tags"
      8 => "custom_fields"
      9 => "created"
      10 => "last_updated"
      11 => "site_count"
      12 => "depth"
    ]
    "hidden" => []
    "visible" => []
  ]
  #extraColumns: []
  #totalRecords: null
  #filteredRecords: null
  #autoFilter: true
  #filterCallback: null
  #templates: array:4 [▼
    "DT_RowId" => ""
    "DT_RowClass" => ""
    "DT_RowData" => array:377 [▼
      95 => array:13 [▼
        "id" => 95
        "url" => "https://myapiendpoint.dev"
        "display" => "Idaho Falls"
        "name" => "Idaho Falls"
        "slug" => "idaho-falls"
        "parent" => null
        "description" => ""
        "tags" => []
        "custom_fields" => []
        "created" => "2023-12-08T04:49:35.371160Z"
        "last_updated" => "2023-12-08T04:49:35.371181Z"
        "site_count" => 0
        "_depth" => 0
      ]
      98 => array:13 [▶]
	  ...
	  1564 => array:13 [▶]
    ]
    "DT_RowAttr" => []
  ]
  #orderCallback: null
  #skipPaging: false
  #appends: []
  #config: 
Yajra\DataTables\Utilities\Config {#1645 ▼
    -repository: 
Illuminate\Config\Repository {#55 ▼
      #items: array:26 [ …26]
    }
  }
  #serializer: ? mixed
  #searchPanes: []
  #transformer: ? mixed
  #editOnlySelectedColumns: false
  +original: 
Illuminate\Support\Collection {#1352 ▶}
  -offset: 0
  +collection: 
Illuminate\Support\Collection {#1352 ▼
    #items: array:10 [▼
      "data" => 
Illuminate\Support\Collection {#1632 …2}
      "count" => 377
      "next" => null
      "previous" => null
      "draw" => 1
      "recordsTotal" => 377
      "recordsFiltered" => 377
      "offset" => 0
      "length" => 10
      "input" => ""
    ]
    #escapeWhenCastingToString: false
  }
}
```
So, my question @yajra is this: why is the columns array 377 items long? and is this affecting the load of the data? Regardless, if I _don't_ add the individual rows using addRowData(), as I did here, the result is the same. The data doesn't load on 1st draw, and if I change the number of items shown, _then_ I see all my items in the DT (even though paging doesn't work for some reason...??). Also, I am getting an error on 1st load in the console: "Uncaught TypeError: Cannot read properties of undefined (reading '_select')"

I can't seem to get past these two problems. Any help would be greatly appreciated! :-)

**Reply by yajra** [reply]
Since you are using server-side:

1. No need to pass the data

```blade
<div class="card-body">
  {{ $dataTable->table() }}
</div>

@push('scripts')
{{ $dataTable->scripts() }}
@endpush
```

2. You can define the data source inside the query method:

```php
public function query()
{
  $tokens = ['offset' => request('start') ?? '0'];
  $response = Netbox::getData('dcim/regions?limit=0', $tokens);

  return $response->get('results');
}
```

3. On your dataTable method, the result from the `query` method is injected as `$query`

```php
public function dataTable($query)
{
  return (new CollectionDataTable($query)
    ->addColumn(...) // etc
    ->setRowId('id');
}
```

**Reply by yajra** [reply]
Concerning https://github.com/yajra/laravel-datatables/issues/1987#issuecomment-2161045853, you might also need to use `skipPaging`

```php
public function dataTable($query)
{
  return (new CollectionDataTable($query)
    ->skipPaging()
    ->setRowId('id');
}
```

**Reply by bigdogdman** [reply]
I'm sorry, I'm just..._super_ confused now. I'm really struggling here....
Thank you for your response. I have already implemented your first suggestion, and I think I've already botched #'s 2 & 3. Let me give you more of my code so you can see what I'm working with. The code below "works". I get the DT to show up, and I get the data to show up after ajax.reload().

So here is my 'query' method:
```
    public static function getRegions(RegionsDataTable $dataTable)
    {
        $response = Netbox::getData('dcim/regions?limit=0', [], true);
        $data = DataTables::of(collect($response));
        return $dataTable->render('layouts.index', compact('data'), ['title' => 'REGIONS']);
    }
```

Here is my getData method return:
```
        return collect(collect($response->json())->get('results'));
        return collect($data);
```
Oh wait, _that's_ what you're saying in # 2...uhh, okay so I guess getData would then be my 'query' method.

Ok so here is my 'dataTable' method (I think?):
```
    public static function getRegionData(Request $request)
    {
        // Capture query parameters
        $queryParams = $request->query();
        $queryParams['offset'] = $queryParams['start'] ?? '0';
        $tokens = $queryParams;
        $response = Netbox::getData('dcim/regions?limit=0', $tokens);
        $data = new CollectionDataTable($response)->toJson();   // without the ->toJson() at the end, it blows up (datatype mismatch)
        return $data;
    }

```
And here is my DataTable:
```
    public function html(): HtmlBuilder
    {
        return $this->builder()
                    ->setTableId('regions-table')
                    ->columns($this->getColumns())
                    ->minifiedAjax()
                    ->ajax(route('region-data'))
                    //*/
                    //->dom('Bfrtip')
                    ->orderBy([0, 'asc'])  // dynamic??
                    ->paging()
                    ->serverSide();

                    //->selectStyleSingle()
                    //->selectSelector('td:first-child')
                    /* ->buttons([
                        Button::make('excel'),
                        Button::make('csv'),
                        Button::make('pdf'),
                        Button::make('print'),
                        Button::make('reset'),
                        Button::make('reload')
                    ])
                    ->drawCallbackWithLivewire() // not working for me...
                    ->orderBy(3, 'asc')
                    ->selectStyleOS()
                    ->selectSelector('td:first-child')
                    ->buttons([
                        Button::make(),
                        //CreateButton::make(),
                        //EditButton::make(),
                        //RemoveButton::make(),
                        //ReloadButton::make(),
                    ])
                    ->editors([         // this seems to break my DataTable
                        Editor::make()
                              ->fields([
                                  Field::make('video_id'),
                                  Text::make('title'),
                                  Select2::make('tags')
                                         ->data('tags[].name')
                                         ->label('Tags')
                                         ->opts([
                                             'multiple' => true,
                                         ])
                                         //->modelOptions(Netbox::class, 'name', 'name'),   // ??
                              ]),
                    ]);
                    */
    }

    /**
     * Get the dataTable columns definition.
     */
    public function getColumns(): array
    {
        $cssClasses = ['class' => 'text-center text-white'];
        return [
            Column::make('id')->width(60)->attributes($cssClasses),
            Column::make('name')->width(240)->attributes($cssClasses),
            //Column::make('url')->width(500)->attributes($cssClasses),   // something special for this? redirect to /regions/{id}, maybe???
            Column::make('display')->width(240)->attributes($cssClasses),
            //Column::make('description')->width(300)->attributes($cssClasses),
            Column::make('created')->width(260)->attributes($cssClasses),
            Column::make('last updated')->data('last_updated')->width(260)->attributes($cssClasses),

            //Column::computed('action')->exportable(false)->printable(false)->width(60)->attributes($cssClasses),
            //Column::make('site_count'),
            //Column::make('slug'),
            //Column::make('parent'),
            //Column::make('depth')->data('_depth'),
            //Column::make('tags'),           //"tags" => [] -> will need special processing if used
            //Column::make('custom_fields'),  //"custom_fields" => [] -> will need special processing if used
        ];
    }
```

And here is my index:
```
<x-app-layout>
    <x-slot name="header">
        <x-footprint-menu />
    </x-slot>
    <x-slot name="slot">
        <div class="container text-white">
            <div class="card">
                <div class="card-header">{{ $title }}</div>
                <div class="card-body">
                    {{ $dataTable->table() }}
                </div>
            </div>
        </div>
        @push('scripts')
            {{ $dataTable->scripts() }}
        <script>
            setTimeout(function () {
                var table = $('#{{ strtolower($title) }}-table').DataTable();
                table.ajax.reload();
            }, 500);
        </script>
        @endpush
    </x-slot>
</x-app-layout>
```

So now that I've inundated you w/all _that_, I am so terribly sorry. I've just been pulling my hair out over this. Ever since the very beginning, I've been getting an error in the initial DataTable load: `Uncaught TypeError: Cannot read properties of undefined (reading '_select')` It's not loading the data into my DataTable initially, thus the `table.ajax.reload();`.

So I am trying to understand/implement your directions, particularly # 2. You say I can do `return (new CollectionDataTable($query)->addColumn(...)->setRowId('id');`? I have been wondering where these directives belonged! Because no matter where I put them, they didn't seem to do anything useful! So can you please help me w/the rest of these directives (or direct me to your docs)? How to get them set up properly? Because I'm now positive I've been misconfiguring stuff from the start, particularly w/the `action` column. I can't get buttons to generate, or the Action column to even work...which is why it's commented out in my DataTable.

Okay so now I have 2 problems: My data doesn't display on initial load of the DT, and trying to implement #'s 2 & 3. Ok make that 3 problems. I'm also trying to link stuff in my DT to other API endpoints (dynamically) but everything I've found online has been very...not helpful w/this. Part of my $data object is a url index that I'm wanting to attach to either rows or just the Id values in the DT; would I do that in the (new CollectionDataTable()) section?

Also, why is it blowing up if I don't have the ->toJson at the end? You said in # 3 I didn't need that? Well, I guess you didn't say _that_; is it ok if that's the last one on the chain though?

...........Oh my goodness I am a complete idiot. Please give me one second to implement your changes...
Okay so now it says this:
![Screenshot 2024-06-13 093723](https://github.com/yajra/laravel-datatables/assets/5806390/00be1d27-90bf-4dc3-9abf-4fe9903a5069)
I took out the QueryBuilder type and it still gave that warning.

But I ran it anyway, and everything worked _exactly_ the same as before. Same console error for _select on initial load, and same (successful) reload from my setTimeout():
![Screenshot 2024-06-13 080454](https://github.com/yajra/laravel-datatables/assets/5806390/9d93e018-4a1b-40d2-ab19-4e2f0aa6e5c7)

Oh, and here is my _new_ DataTable:
```
class RegionsDataTable extends DataTable
{
    /**
     * Build the DataTable class.
     *
     * @param QueryBuilder $query Results from query() method.
     */
    public function dataTable(QueryBuilder $query)
    {
        return (new CollectionDataTable($query))
            ->addColumn('action', '<x-button>TEST</x-button>')
            ->skipPaging()
            ->setRowId('id');
    }

    /**
     * Get the query source of dataTable.
     */
    public function query(Netbox $model): QueryBuilder
    {
        // Capture query parameters
        $queryParams = Request()->query();
        $queryParams['offset'] = $queryParams['start'] ?? '0';
        //dd($queryParams);
        $tokens = $queryParams;
        //$tokens = ['offset' => request('start') ?? '0'];
        $response = $model->getData('dcim/regions?limit=0', $tokens);
        
        return $response->get('results');      
    }

    /**
     * Optional method if you want to use the html builder.
     */
    public function html(): HtmlBuilder
    {
        return $this->builder()
        ...all the rest is the same.
```

I'm assuming that this line in my getRegions call is very important (and not being called correctly, apparently): `return $dataTable->render('layouts.index', compact('data'), ['title' => 'REGIONS']);`
Any changes to any of the code (_not_ inside my RegionsDataTable) seems to break everything...
Okay, I don't seem to be making any more headway, so I'll stop here.
So....I'm _clearly_ doing _many_ things wrong....Help please??  :`(

**Reply by bigdogdman** [reply]
oh, and I didn't even tell you the best part! Here's what it's doing _without_ the setTimeout:
![Screenshot 2024-06-13 080245](https://github.com/yajra/laravel-datatables/assets/5806390/1e15f7f3-8249-42c3-8bfe-22472394195e)
Isn't that awesome?!! :``(
Oh, sorry, the _url error doesn't seem to be happening anymore, so you can ignore that one. :")

**Reply by yajra** [reply]
The query method should return an array or better a collection. Please take a look at the adjusted code below.

```php
    public function dataTable(Collection $query): CollectionDataTable
    {
        return (new CollectionDataTable($query))
            ->addColumn('action', '<x-button>TEST</x-button>')
            ->skipPaging()
            ->setRowId('id');
    }

    public function query(Netbox $model): Collection
    {
        // Capture query parameters
        $queryParams = Request()->query();
        $queryParams['offset'] = $queryParams['start'] ?? '0';
        //dd($queryParams);
        $tokens = $queryParams;
        //$tokens = ['offset' => request('start') ?? '0'];
        $response = $model->getData('dcim/regions?limit=0', $tokens);
        
        return collect($response->get('results')); // added collect but remove if it's a collection already.   
    }

**Reply by bigdogdman** [reply]
Thank you for that correction, but I'm hitting neither the dataTable nor the query functions. Clearly I'm doing something wrong. I went back through your quick-starter guide to make sure I had everything set up correctly, and basically the only steps I didn't follow precisely were 3 & 6. I also corrected my getRegions method to look like this: `return $dataTable->render('layouts.index');` but it didn't fix anything. Should my callback route look different? the getRegionData? That's not making use of the RegionsDataTable object. Would that be where I'd call the query function? Because if I call it from there, with just a `return $dataTable->query();` it shows no rows in the DT; it doesn't appear to be returning the correct datatype (afaik). It's not blowing up, it's just not returning the collection object that the DT is expecting. It appears to be a collection of _just_ the rows of data, without the rest of the pieces, like draw & count, etc. Using this line: `$data = (new CollectionDataTable($response))->toArray();` and then returning `collect($data)` appears to be working; however, it's still not displaying any rows of data (or the pagination footer) on initial load. I also still don't appear to be hitting RegionsDataTable::dataTable()...

**Reply by yajra** [reply]
The controller can be a simple get:

```php
Route::get('/regions', [RegiosController::class, 'index']);

-- regions controller
public function index(RegionsDataTable $dataTable)
{
  return $dataTable->render('regions.index');
}
```

**Reply by yajra** [reply]
Also, enable `APP_DEBUG=true` then inspect the network request. The ajax response should give you some hints on the issue.

**Reply by bigdogdman** [reply]
So you're saying that the ajax callback should _also_ be hitting index? And, which ajax response should I look at after turning debug on? The callback response? Because all I'm seeing from that one is the data that's displayed in the table. This is the full backtrace of the error msg in the console:
![image](https://github.com/yajra/laravel-datatables/assets/5806390/e202aa69-eedb-4674-a934-a7144560c312)

Ok so after I switched the callback to hit index, I was finally hitting RegionsDataTable::dataTable($query) method, but then that threw this error: `Yajra\\DataTables\\Processors\\DataProcessor::editColumns(): Argument #2 ($row) must be of type object|array, int given, called in vendor/yajra/laravel-datatables-oracle/src/Processors/DataProcessor.php on line 75`.
Is that the clue we were after? Cuz I've definitely seen that one before...also, it's still not displaying the data on initial load (as it's still giving the _select error).

Also, switching the callback to hit index ('/regions'), the data object changed from a Collection (from query()) to a CollectionDataTable (from dataTable()). Collection return object works for my RegionsDataTable, CollectionDataTable doesn't seem to. If that helps at all.

**Reply by yajra** [reply]
Can you create a repository to replicate the issue?

**Reply by bigdogdman** [reply]
I...cannot. I am limited by my work and am unable to make any of this public. It wouldn't work even if I could. It's behind a login _and_ a VPN. You can see more of my history with this problem [here](https://datatables.net/forums/discussion/79206/datatable-being-wrapped-in-another-datatable-trying-to-display-data-on-load/p1). Hopefully this can shed some more light on what could possibly be happening.
I apologize for taking up so much of your time on this issue, but you are literally my last hope on getting this working. I will do anything I possibly can to help you diagnose this problem short of making this project publicly viewable.
